### PR TITLE
refactor(v5.0): TableSpec — one source of truth per table (Tier 3.2)

### PR DESCRIPTION
**Pull request by Papamzor**
## Tier 3.2 — TableSpec: one source of truth per table

Stacked PR **2 of 3** for v5.0 "Boron" Tier 3. **Base: `feature/v5.0_response-contract-rest`** (merge PR 1 first).

Per-table config was scattered across five collections in `constants.py`. `cf2d7e2` needed coordinated edits to three of them for one logical change, and nothing forced them to describe the same table set.

New `table_spec.TABLE_SPECS` collapses it into one `TableSpec` per table. `constants.py` now **derives** the five collections (`TABLE_CONFIGS`, `ESSENTIAL_FIELDS`, `DETAIL_FIELDS`, `TABLE_ERROR_MESSAGES`, `TABLES_WITHOUT_RECORD_IDENTITY` + text-search map) from it — same shapes, same keys, **byte-identical values (verified against HEAD)**. Every `from constants import ...` consumer is unchanged.

The `task_sla` record-identity foot-gun is now **structural**: `number_field=None` means "no record number", and `TABLES_WITHOUT_RECORD_IDENTITY` is derived as the set of number_field-less specs — a table with no number field cannot be left out of the guard.

NEW `tests/test_table_spec.py` pins the key-set consistency the scattered dicts never had, the derived identity guard, and per-spec well-formedness.

**Suite: 1117 passed.** No behavior change; no consumer edits.

### CODE REVIEW

File: constants.py [hunk 1]
```diff
@@ -4,6 +4,13 @@
 
 import os
 
+# Per-table configuration is now defined once, in table_spec.TABLE_SPECS (v5.0
+# "Boron" Tier 3.2). The collections below (TABLE_CONFIGS, ESSENTIAL_FIELDS,
+# DETAIL_FIELDS, TABLE_ERROR_MESSAGES, TABLES_WITHOUT_RECORD_IDENTITY, the
+# text-search map) are DERIVED from it so existing `from constants import ...`
+# consumers are unaffected. Change a table there, not here.
+from table_spec import TABLE_SPECS, TableSpec
```
Comment (Papamzor) [inline constants.py]: **[bug]** `constants.py` now hard-depends on root module `table_spec` (`from table_spec import TABLE_SPECS, TableSpec`), but this PR does not add `table_spec.py` to `scripts/build_mcpb.py` `ROOT_FILES`. That whitelist is the only set of root Python files copied into `dist/mcpb-staging/` for the Claude Desktop Extension. An mcpb build after this PR will package a `constants.py` that fails at import with `ModuleNotFoundError: No module named 'table_spec'`, breaking every tool path that touches constants. Existing packaging tests (`tests/test_pyproject_sync.py`) do not assert that every root import of a staged module is also staged.

**Suggestion:** Add `"table_spec.py"` to `ROOT_FILES` in `scripts/build_mcpb.py`, and add a regression test that either (a) imports the staged package tree, or (b) asserts every non-stdlib root import of a file in `ROOT_FILES` is itself listed in `ROOT_FILES` / package dirs.

File: tests/test_table_spec.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import constants as c
+from table_spec import TABLE_SPECS, TableSpec
+
+_SPEC_KEYS = set(TABLE_SPECS)
+
+
+class TestKeySetConsistency:
+    """Every derived view covers exactly the TABLE_SPECS table set."""
+
+    def test_derived_dicts_share_the_spec_key_set(self):
```
Comment (Papamzor) [inline tests/test_table_spec.py:22]: **[suggestion]** `test_derived_dicts_share_the_spec_key_set` only asserts that `TABLE_CONFIGS`, `ESSENTIAL_FIELDS`, `DETAIL_FIELDS`, and `TABLE_ERROR_MESSAGES` share the `TABLE_SPECS` key set. Nothing asserts that derivation maps the correct attributes — e.g. `ESSENTIAL_FIELDS[name] == list(spec.essential_fields)`, `TABLE_ERROR_MESSAGES[name] == spec.error_message`, or each `TABLE_CONFIGS[name]` field equals the matching `TableSpec` attribute. A wrong attribute wiring in `constants.py` (e.g. `error_message` → `display_name`, swapped essential/detail, or wrong `supports_*` field) would pass these tests while breaking consumers. The PR claim of byte-identical values vs HEAD is therefore not locked by this suite.

**Suggestion:** Add a per-table derivation fidelity test that walks `TABLE_SPECS` and asserts each derived collection’s values (and `TABLE_CONFIGS` keys including `api_name == name`) equal the corresponding spec fields. Optionally also assert `text_search_field_for(name) == spec.text_search_field` and `TEXT_SEARCH_FIELD_BY_TABLE` equals the set of non-default overrides.

File: table_spec.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    state_field: str
+    text_search_field: str
+    supports_work_notes: bool
+    supports_comments: bool
+    essential_fields: Tuple[str, ...]
+    detail_fields: Tuple[str, ...]
+    error_message: str
+
+    @property
+    def has_record_identity(self) -> bool:
```
Comment (Papamzor) [inline table_spec.py:66]: **[suggestion]** `has_record_identity` (and thus `TABLES_WITHOUT_RECORD_IDENTITY`) is keyed only on `number_field is not None`. There is no structural invariant that `number_prefix is None` iff `number_field is None`. Consumers still read `TABLE_CONFIGS[t]["number_prefix"]` for prefix logic; a future edit can set `number_prefix="X"` with `number_field=None` (guard refuses identity tools but prefix looks real) or `number_prefix=None` with `number_field="number"` (identity tools allowed, prefix consumers see None). The old hand list was implicitly tied to “no number / no prefix”; the new model splits those without enforcing co-movement. Tests only assert `number_field` for identity, not prefix alignment.

**Suggestion:** Add `__post_init__` (or a registry validation test) that requires `(number_field is None) == (number_prefix is None)`, and for identity tables assert `number_field == "number"` and a non-empty prefix. Keep `has_record_identity` as the guard source of truth.

File: table_spec.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+well-formed.
+"""
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Optional, Tuple
+
+# Default free-text search column, and the task_sla dot-walk override. task_sla
+# carries no short_description of its own — the text lives on the referenced task.
+_DEFAULT_TEXT_SEARCH_FIELD = "short_description"
```
Comment (Papamzor) [inline table_spec.py]: **[suggestion]** Free-text field strings are duplicated outside the SSOT story: `table_spec` owns `_DEFAULT_TEXT_SEARCH_FIELD` / `_TASK_SLA_TEXT_SEARCH_FIELD`, while `constants` still defines public `TEXT_SEARCH_FIELD` / `TASK_SLA_TEXT_SEARCH_FIELD` as independent string literals. `TEXT_SEARCH_FIELD_BY_TABLE` filters with `!= TEXT_SEARCH_FIELD` (constants), while specs default via the private `_DEFAULT_*` constant. `test_text_search_field_default_or_dotwalk` only requires `"." in` the task_sla field, not equality with `c.TASK_SLA_TEXT_SEARCH_FIELD`. Drift between the two pairs weakens the “one place” claim for the exact columns that historically caused silent ServiceNow drops.

**Suggestion:** Derive the public constants from the same literals used by specs (e.g. export the defaults from `table_spec`, or set `TEXT_SEARCH_FIELD` / `TASK_SLA_TEXT_SEARCH_FIELD` from `TABLE_SPECS` / shared module-level constants). Tighten the task_sla test to assert the exact `"task.short_description"` value (or `c.TASK_SLA_TEXT_SEARCH_FIELD`).

File: constants.py (hunk 1 above)
Comment (Papamzor) [inline constants.py]: **[nit]** `TableSpec` is imported in `constants.py` but never referenced there; only `TABLE_SPECS` is used for derivation. Dead import noise (and a slightly broader import surface than needed).

**Suggestion:** Change to `from table_spec import TABLE_SPECS` unless a type annotation / re-export is planned in this PR.

File: table_spec.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+def _spec(**kwargs) -> TableSpec:
+    # Every table addressable by number uses the literal `number` field, and
+    # short_description as its text-search column; the two exceptions (task_sla)
+    # pass explicit values. Defaults keep the registry below readable.
+    kwargs.setdefault("number_field", "number")
+    kwargs.setdefault("text_search_field", _DEFAULT_TEXT_SEARCH_FIELD)
+    return TableSpec(**kwargs)
+
+
+TABLE_SPECS = {
```
Comment (Papamzor) [inline table_spec.py]: **[nit]** Individual `TableSpec` instances are `@dataclass(frozen=True)` with tuple field lists (good), but `TABLE_SPECS` itself is a plain mutable dict. Accidental `TABLE_SPECS[...] = ...` / `.pop` after import would desync from already-materialized derived views in `constants` (those are snapshotted at import). Derived `ESSENTIAL_FIELDS` / `DETAIL_FIELDS` / `TABLE_CONFIGS` remain fully mutable by design for historical public types — acceptable — but the SSOT registry is the one structure that most benefits from immutability.

**Suggestion:** Wrap the registry in `types.MappingProxyType({...})` (or assign to a private `_TABLE_SPECS` and expose a read-only view) so only process-level reassignment of the name can change it. Leave derived consumer dicts mutable if callers depend on that.

**Comment by Papamzor** [comment]
Addressed in dc66948:
- **Packaging bug**: added `table_spec.py` to `scripts/build_mcpb.py` ROOT_FILES, and a new `test_every_root_import_of_a_local_module_is_staged` that scans every staged root .py's imports and fails if any repo-local module it imports is not itself staged (verified it flags `table_spec` on the old list).
- `TableSpec.__post_init__` enforces the number_field/number_prefix co-movement invariant (both set or both None; identity tables use `number_field=="number"` + non-empty prefix).
- Deduped the free-text-search literals: table_spec exposes `DEFAULT_TEXT_SEARCH_FIELD`/`TASK_SLA_TEXT_SEARCH_FIELD` and constants derives from them; dropped the unused `TableSpec` import.
- `TABLE_SPECS` wrapped in `MappingProxyType` (read-only SSOT).
- Added per-table derivation-fidelity tests (values, not just key sets) + tightened the task_sla text-search assertion to the exact value.
Suite 1126 → green.
